django/db/models/query_utils.py

```diff
@@ -73,6 +73,8 @@ def __copy__(self):
         obj.__dict__ = self.__dict__.copy()
         return obj
 
+    copy = __copy__
```

Comment:
If we implement this on `Node`, we can simplify the following to assume that we always have `.copy()`:

https://github.com/django/django/blob/fd999318ad78613227cdb7c5656345d9e216802b/django/db/models/query_utils.py#L48

Reply:
Hmmmm, maybe. I was trying to avoid hoisting things up to `Node` to keep the changes self contained and easily evaluated, mostly so I didn't need to keep the definitions in `WhereNode` in my head too. 

Thinking it through now I've done the changes and you've suggested it, the shallow copy of a `WhereNode` is already technically problematic. (I have my eyes on opening a ticket at some juncture), and wouldn't become any more so by having a custom one defined on `Node`, and `WhereNode` already rebinds `copy` to `clone` so I think it _would_ be fine?

Reply:
> ..., and `WhereNode` already rebinds `copy` to `clone` so I think it _would_ be fine?

Indeed. As `WhereNode` implements `.copy()` to call `.clone()` this is fine. In fact, removal of this branch that relies on `copy.copy()` makes it even better as it reduces the risks. I guess the question is whether we even need `Node.__copy__()` in that case, or whether we can roll it directly into `Node.copy()`?

Reply:
_Personally_, I'm not a fan of the methods being `clone` and `copy` when the copy & deepcopy protocols already exist. I have no idea how deep or shallow to expect them to be :)
For consistency, I'd argue that it should be implemented as `__copy__` to match the concept of `Node.__deepcopy__`, and rebindings to `copy` and `clone` would only exist for API compatibility.

Reply:
So, if you'll indulge me now I've got a bit of time to write this down, I'm going to outline why I want to move to `__copy__` here, because I don't know the etiquette of opening pull requests just to see if the CI suite fails, for things which don't have a ticket and may not be appropriate for one.

Though we're talking here about Node & Q, it's cloning `WhereNode` which I'm thinking about, whose cloning is also expensive (line profiling `Query.clone` suggests), but it needs to be done you can't just copy the `__dict__` value because attributes get mutated onto the instance later in the lifecycle (eg: `contains_aggregate` turns up once the cached property is called, other things turn up during building out the SQL to execute IIRC, etc.), so it recursively calls .clone() if the child node has it.

My (half-baked) **working theory** is, though, that this is only necessary _because_ it isn't clear how shallow or deep the copy is, and the use of `copy` and `clone` are not _necessarily_ consistent vs having `__copy__` and `__deepcopy__`

For example, locally the test suite completes if I implement `WhereNode.clone` as:
```
clone = self._new_instance(children=None, connector=self.connector, negated=self.negated)
clone.children = self.children[:]
return clone
```
instead of:
```
clone = self.__class__._new_instance(children=None, connector=self.connector, negated=self.negated)
clone.children = self.children[:]
for child in self.children:
    if hasattr(child, 'clone'):
        clone.children.append(child.clone())
    else:
        clone.children.append(child)
```
Now if that were to hold true for the bits I can't test locally (eg: things like `GDALRaster` getting shallow copied, or whatever else _might_ have clone currently), then the `Node.__copy__` method could **not** copy `__dict__`, and _just_ copy `children`, and then it functionally returns the same data as `__deepcopy__` (new instance + children only), the differences are clear, and both WhereNode and Q can just have `copy & clone = __copy__` and be simpler & I _think_ faster (additionally re-binding is _slightly_ faster than defining `def copy(self): return self.clone()` AFAIK)

Then the whole implementation would be, in one place:
```
def __copy__(self):
    x = self._new_instance(...)
    x.children = self.children[:]
    return x

def __deepcopy__(self, memo):
    x = self._new_instance(...)
    x.children = copy.deepcopy(self.children, memo)
    return x

copy = __copy__
clone = __copy__
```